RunScript.c

```diff
@@ -0,0 +1,153 @@
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "Action.h"
+#include "InfoScreen.h"
+#include "MainPanel.h"
+#include "Object.h"
+#include "Panel.h"
+#include "Process.h"
+#include "Row.h"
+#include "ScriptOutputScreen.h"
+#include "Settings.h"
+#include "XUtils.h"
+
+
+static void write_row(Row* row, int write_fd) {
+   Process* this = (Process*) row;
+   assert(Object_isA((const Object*) this, (const ObjectClass*) &Process_class));
+
+   int pid_length = snprintf(NULL, 0, "%d", row->id);
+   char pid[pid_length + 1];
+   snprintf(pid, pid_length + 1, "%d", row->id);
+
+   char* pid_str = String_cat(pid, "\t");
+   char* user = String_cat(this->user, "\t");
+   char* cmd = String_cat(Process_getCommand(this), "\n");
+   char* user_and_cmd = String_cat(user, cmd);
+   char* line = String_cat(pid_str, user_and_cmd);
+
+   // writes PID\tUser\tCommand\n, using TSV format
+   size_t count = strlen(line);
+   char* line_start = line;
+   while (count > 0) {
+      ssize_t res = write(write_fd, line_start, strlen(line));
+      if ((res == -1 && errno != EINTR) || res == 0)
+         break;
+
+      count -= res;
+      line_start += res;
+   }
+
+   free(pid_str);
+   free(user);
+   free(cmd);
+   free(user_and_cmd);
+   free(line);
+}
+
+void RunScript(State* st) {
+   int child_read[2] = {0, 0};
+   int child_write[2] = {0, 0};
+
+   if (pipe(child_read) == -1)
+      return;
+   if (pipe(child_write) == -1) {
+      close(child_read[0]);
+      close(child_read[1]);
+      return;
+   }
+
+   pid_t child = fork();
+   if (child == -1) {
+      close(child_read[0]);
+      close(child_read[1]);
+      close(child_write[0]);
+      close(child_write[1]);
+      fprintf(stderr, "fork failed\n");
+      return;
+   } else if (child == 0) {
+      close(child_read[1]);
+      dup2(child_read[0], STDIN_FILENO);
+      close(child_read[0]);
+
+      close(child_write[0]);
+      dup2(child_write[1], STDOUT_FILENO);
+      dup2(child_write[1], STDERR_FILENO);
+      close(child_write[1]);
+
+      char* home = getenv("XDG_CONFIG_HOME");
+      if (!home)
+         home = String_cat(getenv("HOME"), "./config");
+
+      const char* path = String_cat(home, "/htop/run_script");
+      FILE* file = fopen(path, "r");
+      if (file) {
+         execl(path, path, NULL);
+         // should not reach here unless execl fails
+         fprintf(stderr, "error excuting %s\n", path);
+         perror("execl");
+      } else {
+         // check if htoprc has something
+         const char* htoprc_path = st->host->settings->scriptLocation;
+         execl(htoprc_path, htoprc_path, NULL);
+
+         // only reach here if execl fails
+         fprintf(stderr, "error executing %s from htoprc", htoprc_path);
+         fprintf(stderr, "if you expected your runscript to be executed, htop looked for it at %s", path);
+         perror("execl");
+      }
+      exit(1);
```

Comment:
IIRC this is inconsistent with other locations that exec where return on error is 127.